metricplugin/tcp_client.go

```diff
@@ -0,0 +1,403 @@
+
+		buf := pool.GlobalPool.Get(1024 * 512)
+		bbuf := bytes.NewBuffer(buf)
+		bbuf.Reset()
+		w := json.NewEncoder(bbuf)
+		if err := w.Encode(msg); err != nil {
+			panic(fmt.Sprintf("encode %s: unable to marshal %+v to JSON: %s", c.remote, msg, err))
+		}
+
+		log.Debugf("encode %s: encoded outgoing message: %q", c.remote, bbuf.String())
+
+		select {
+		case encodedOut <- bbuf.Bytes():
+		case <-c.closed:
+			log.Debugf("encode %s: client closed", c.remote)
+			return
+		}
+	}
+}
+
+// Asynchronous loop to compress outgoing messages.
+// This takes JSON-encoded messages and compresses them using gzip.
+// For each message, a buffer is taken from the global buffer pool.
+// The buffer for the incoming, JSON-encoded bytes is returned to the global
+// pool after use.
+// The consumer of compressedOut should return the buffer for the compressed
+// message to the global pool.
+func (c *tcpClient) compressLoop(encodedIn <-chan []byte, compressedOut chan []byte) {
+	defer func() {
+		log.Debugf("compress %s: exiting", c.remote)
+		close(compressedOut)
+	}()
+
+	preCompressionCounter := wiretapSentBytes.With(prometheus.Labels{"compressed": "false"})
+	postCompressionCounter := wiretapSentBytes.With(prometheus.Labels{"compressed": "true"})
+
+	var msgBytes []byte
+	buf := make([]byte, 0)
+	bbuf := bytes.NewBuffer(buf)
+	zw := gzip.NewWriter(bbuf)
+
+	for {
+		select {
+		case msgBytes = <-encodedIn:
+		case <-c.closed:
+			log.Debugf("compress %s: client closed", c.remote)
+			return
+		}
+
+		log.Debugf("compress %s: compressing %d bytes: %q", c.remote, len(msgBytes), msgBytes)
+
+		// Get a buffer, re-initialize the gzip writer on that.
+		buf = pool.GlobalPool.Get(1024 * 512)
+		bbuf = bytes.NewBuffer(buf)
+		bbuf.Reset()
+		zw.Reset(bbuf)
+
+		_, err := zw.Write(msgBytes)
+		if err != nil {
+			panic(fmt.Sprintf("compress %s: unable to compress: %s", c.remote, err))
+		}
+		if err = zw.Close(); err != nil {
+			panic(fmt.Sprintf("compress %s: unable to compress: %s", c.remote, err))
+		}
+
+		log.Debugf("compress %s: compressed from %d to %d bytes: %s", c.remote, len(msgBytes), bbuf.Len(), hex.EncodeToString(bbuf.Bytes()))
+
+		// Record in Prometheus.
+		preCompressionCounter.Add(float64(len(msgBytes)))
+		postCompressionCounter.Add(float64(bbuf.Len()))
+
+		// Release the message byte buffer from the encoder.
+		pool.GlobalPool.Put(msgBytes)
+
+		select {
+		case compressedOut <- bbuf.Bytes():
+		case <-c.closed:
+			log.Debugf("compress %s: client closed", c.remote)
+			return
+		}
+	}
+}
+
+var (
+	// ErrTimeout is returned if a send operation times out.
+	ErrTimeout = errors.New("timeout sending")
+	// ErrClosed is returned for method calls on a closed or closing client.
+	ErrClosed = errors.New("client closed")
+	// ErrBackpressure is returned if the outgoing buffer of messages is full.
+	ErrBackpressure = errors.New("dropped due to backpressure")
+)
+
+// Asynchronous loop to send framed messages to the client.
```

Comment:
Any particular reason why enconding, compressing and sending are all handled in their own async loop instead of one loop for a combination of all tasks?